site/content/concepts/command-and-control.md

```diff
@@ -3,29 +3,74 @@ title = "Command & Control"
 weight = 196
 +++
 
-*Business applications* can send commands to devices. The intended cases as well the state of implementation are explained here. 
+Since Hono 0.6 *Business applications* can send commands to devices following the [Command & Control API]({{< relref "api/Command-And-Control-API.md" >}}). This concept describes how to use this API to send commands to devices connected to Hono via the Adapters. These direct commands follow a request-response pattern and expect an immediate confirmation of their result.  
  
 <!--more-->
 
-This concept refers to the first version of Command and Control in Hono 0.6. With new versions this will be extended.
-
-## Sending a command
-
-**Use Cases**
-
-1. The application sends a command to a Device and expects a direct response from the device.
-2. The application sends a command, but does not need a direct response. Maybe it gets later an normal event indicating a response.
+This concept refers to the first implementation of Command and Control in Hono 0.6. With new versions this will be extended and documentation parts will be also transferred to the User Guide sections (e.g. HTTP Adapter and MQTT Adapter).
 
 ## Command & Control over HTTP Adapter
 
-The following sequence diagram gives an overview of case 1. with a device connected via HTTP.
+The following sequence diagram gives an overview of a device connected via HTTP, which gets a command from the business application as soon as it gets online.
  
 ![Command & Control over HTTP Adapter](../command_control_concept_http.png) 
 
-With the `ttd`-[notification]({{< relref "device-notifications.md" >}}) in (1) the device indicates it will stay connected for max. 30 seconds. This notification will be consumed by the application (2) and it now tries to send a command (3) to the device at the given address `control/TENANT/4711` (If the device would not be connected and the adapter has not established a link for it, the application would got no credits so send the command).
+With the `hono-ttd`-[notification]({{< relref "device-notifications.md" >}}) in (1) the device indicates it will stay connected for max. 30 seconds. This notification will be consumed by the application (2) and it now tries to send a command (3) to the device at the given address `control/TENANT/4711` (If the device is not connected, the adapter has not established a link for it and the application would get no credits so send the command).
 
-The HTTP Adapter gets the command and writes it in the response of the devices `send event` (4).
+The HTTP Adapter gets the command and writes it in the response of the devices `send event` (4). The HTTP Adapter sets the following response headers and optionally a payload.
```

Comment:
does the HTTP adapter only include the command in a response to a successful request (i.e. if the response contains a 2xx status code) or can the command also be included in a response to an unsuccessful request? If so then how does the device know that the response payload is *not* a description of the error that occurred but is a command that it should process?

Reply:
@sophokles73 :+1: IMHO this is a very clever comment/question.

Due to the one-way character of the downstream messages (meaning that there is no answer to them that could be unsuccessful) I see mainly problems with sending the message at all (network problems downstream, QoS level problems with missing acks, ...).

Since currently we can only send a command when a downstream message successfully reached an application, the unsuccessful cases should never have a command for now.

But if we buffer or broker commands in the future, this would change, so this is only an assumption for 0.6 IMHO.

Summarized:
- IMHO we need to decide this question in the future
- but for 0.6 I see no possible occurence of a command in unsuccessful HTTP requests

WDYT?

Reply:
The command should only appear in a successful request. At least for 0.6. But I do not see why it should be in a a response to an unsuccessful request at all? I will add a hint, that the command will only be included in a successful requests response (2xx).